vllm/core/block/cpu_gpu_block_allocator.py

```diff
@@ -204,3 +217,47 @@ def get_common_computed_block_ids(
 
     def all_block_ids(self) -> frozenset[int]:
         return frozenset(self._block_ids_to_allocator.keys())
+
+    def get_seq_swap_out_block_mapping(
+            self, seq: Sequence, block_table: BlockTable,
+            mapping: Dict[Block, Block]) -> BlockTable:
+        # The swap out logic for a sequence, the mapping dict will be updated
+        # and the new block table for swapped out sequence is returned.
+        new_block_table = BlockTable(
+            block_size=self._block_size,
+            block_allocator=self,
+        )
+        for gpu_block in block_table.get_blocks():
+            if gpu_block in mapping:
+                cpu_block = mapping[gpu_block]
+                self.reference(cpu_block)
+            else:
+                cpu_block = new_block_table.allocate(
+                    token_ids=gpu_block.token_ids,
+                    device=Device.CPU,
+                    by_block=True)
+                mapping[gpu_block] = cpu_block
+            self.free(gpu_block)
```

Comment:
This logic strikes me as very similar to per-block forking: https://github.com/vllm-project/vllm/blob/9181552465f165b5f2fac45b3c128be4efc02bcb/vllm/core/block/naive_block.py#L115-L126

Can we refactor so both this and `fork` use the same implementation? might need new API in the block allocator

The downside of not refactoring is that the current abstraction is leaky -- we ideally don't want to share refcounting details between allocators. If we can hide that behind `fork`-like semantics, it decouples the layers.

Reply:
Here I think the semantics  of `reference` is pretty like `fork`, in this setting, we do not need the blocks returned by `fork` and we only want to touch refcounting for a single block instead of all previous blocks in the sequence.